### ISSUE TITLE
Incorrect behavior in symbolic due to poor typing discipline

### ISSUE THREAD

**Issue by PaulGustafson** [tinygrad/tinygrad#2469]
The following test fails
```
    i = Variable("i", 1, 5)
    idx0 = Variable("idx0", 0, i)
    assert idx0 % 2 != idx0
```

This is because get_bounds() can return Nodes, not just integers, which messes up inequalities in the definition of Node.\_\_mod\_\_.  Similar tests fail for MulNodes:

```  
    a = Variable("a", 1, 2)
    u = Variable("u", 1, 2)
    x = MulNode(a,u)
    assert (x < 2) != 0 
```

There are also hacks like this one in `def size()` in shapetracker.py that should be removed:
```while not isinstance(ret, int): ret = ret.max```

Proposed solution: make symbolic.py type safe.

**Comment by PaulGustafson** [comment]
#2474 rules out the first failing test, but the following variant is harder to rule out since many shapetracker unit tests use a similar code path

```  
def test_nested_variable_mod2(self):
    i = Variable("i", 1, 100)
    idx0 = Variable("idx0", 0, i)
    with self.assertRaises(AssertionError):
      assert idx0 % Variable('j', 2, 3) == idx0
```

**Comment by ronak69** [comment]
> Similar tests fail for MulNodes:

this patch fixes the issue. (buf if shapetracker doesn't do `MulNode * MulNode < int` then it does not matter.

```diff
diff --git a/tinygrad/shape/symbolic.py b/tinygrad/shape/symbolic.py
index a9518f12..627bced8 100644
--- a/tinygrad/shape/symbolic.py
+++ b/tinygrad/shape/symbolic.py
@@ -180,7 +180,7 @@ class LtNode(OpNode):
   def __floordiv__(self, b: Union[Node, int], _=False): return (self.a//b) < (self.b//b)
   def get_bounds(self) -> Tuple[int, int]:
     if isinstance(self.b, int):
-      return (1, 1) if self.a.max < self.b else (0, 0) if self.a.min >= self.b else (0, 1)
+      return (1, 1) if (self.a.max < self.b) == True else (0, 0) if (self.a.min >= self.b) == True else (0, 1)
     return (1, 1) if self.a.max < self.b.min else (0, 0) if self.a.min >= self.b.max else (0, 1)
   def substitute(self, var_vals: Dict[VariableOrNum, Node]) -> Node:
     return self.a.substitute(var_vals) < (self.b if isinstance(self.b, int) else self.b.substitute(var_vals))
```

see also: https://github.com/tinygrad/tinygrad/issues/1924#issuecomment-1742153891

**Comment by hristog** [comment]
> this patch fixes the issue. (buf if shapetracker doesn't do `MulNode * MulNode < int` then it does not matter.
> 
> ```diff
> diff --git a/tinygrad/shape/symbolic.py b/tinygrad/shape/symbolic.py
> index a9518f12..627bced8 100644
> --- a/tinygrad/shape/symbolic.py
> +++ b/tinygrad/shape/symbolic.py
> @@ -180,7 +180,7 @@ class LtNode(OpNode):
>    def __floordiv__(self, b: Union[Node, int], _=False): return (self.a//b) < (self.b//b)
>    def get_bounds(self) -> Tuple[int, int]:
>      if isinstance(self.b, int):
> -      return (1, 1) if self.a.max < self.b else (0, 0) if self.a.min >= self.b else (0, 1)
> +      return (1, 1) if (self.a.max < self.b) == True else (0, 0) if (self.a.min >= self.b) == True else (0, 1)
>      return (1, 1) if self.a.max < self.b.min else (0, 0) if self.a.min >= self.b.max else (0, 1)
>    def substitute(self, var_vals: Dict[VariableOrNum, Node]) -> Node:
>      return self.a.substitute(var_vals) < (self.b if isinstance(self.b, int) else self.b.substitute(var_vals))
> ```
> 
> see also: [#1924 (comment)](https://github.com/tinygrad/tinygrad/issues/1924#issuecomment-1742153891)

@ronak69, `== True` is redundant.
On a side note, it's more idiomatic to compare values such as `True`, `False` and `None` via `is` (as opposed to `==`).

**Comment by ronak69** [comment]
@hristog

> `== True` is redundant

it is actually *not*.

it is there to convert a three-state-object (1. `True`, 2. `False` and 3. `LtNode`) into a two-state-object (1. `True` and 2. `False`) which is just a `bool`.

**Comment by PaulGustafson** [comment]
Yes, this is a source of a many annoying bugs.  If you write down a comparator involving Nodes that doesn't immediately reduce to True/False, then it will be a LtNode that gets implicitly cast to True when used as the argument of a conditional.

**Comment by hristog** [comment]
Yeah, my bad - thank you @ronak69, @PaulGustafson! I've updated my comment above, accordingly.
I'll be more careful in the future with reading the context properly first. I confirm we're on the same page, regarding the related type of bugs you've pointed out!

This is still true, though:
> On a side note, it's more idiomatic to compare values such as True, False and None via is (as opposed to ==).

**Comment by chenyuxyz** [comment]
this raises proper Exception now, also most of symbolic is replaced by UOps
